Classes/EventStore/AbstractEventSourcedRepository.php

```diff
@@ -84,28 +78,25 @@ public function findByIdentifier($identifier)
     }
 
     /**
-     * @param  AggregateRootInterface $aggregate
+     * @param AggregateRootInterface $aggregate
+     * @param int $expectedVersion
      * @return void
      */
-    public function save(AggregateRootInterface $aggregate)
+    final public function save(AggregateRootInterface $aggregate, int $expectedVersion = null)
     {
-        $streamName = $this->streamNameResolver->getStreamNameForAggregate($aggregate);
-        try {
-            $stream = $this->eventStore->get($streamName);
-        } catch (EventStreamNotFoundException $e) {
-            $stream = new EventStream();
+        if (!$aggregate instanceof EventSourcedAggregateRootInterface) {
+            throw new \InvalidArgumentException(sprintf('This repository can only save instances implementing EventSourcedAggregateRootInterface, given: "%s"', get_class($aggregate)), 1477226552);
         }
-
+        $streamName = $this->streamNameResolver->getStreamNameForAggregate($aggregate);
         $uncommittedEvents = $aggregate->pullUncommittedEvents();
-        $stream->addEvents(...$uncommittedEvents);
+        $version = $aggregate->getVersion() + 1;
```

Comment:
Hmm, but the internal state of the aggregate at that point is different from the version it was in the stream. That's what I mean with consistent. I mean, in our web-context that is probably not as critical as elsewhere, since no other request will see the aggregate in this inconsistent state - but still anything inside this request could.

If we really have the version as an external measure, I'd rather have us completely keep it off the aggregate. It thereby basically has become the stream version of the aggregate's stream, so why not store and get it from there only rather than expose it to the aggregate?
That would also allow us to enforce "not allow the version to be changed from within the aggregate".